### PR TITLE
[RFC] Make report_error() thread safe

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This PR is supposed to make the whole `report_error()` thing thread safe. The global error buffer is removed. Collecting of errors is done in the frontend. The collected errors are protected by mutexes. Desktop version slightly tested.

### CODE REVIEW

**Comment by janmulder** [comment]
Tried and LGTM, but lets have @dirkhh to have a view to this as well before merging.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: `appendTextToLog` is probably called twice here given `parse_file()` already does a `report_error()`?
do we actually need the extra error string here?

also the `parse_file()` errors strings are translated.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: AFAIK, it's in the same thread, but do we know for sure that `m_lastError` will be related to `loadDivesWithValidCredentials()` at all times here and not being an error for something else?

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: similar to the above Q.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [reply mobile-widgets/qmlmanager.cpp]: similar comment applies to other usages of `m_lastError`.

File: mobile-widgets/qmlmanager.cpp
Comment (bstoeger) [reply mobile-widgets/qmlmanager.cpp]: Yes, re-logging the error message is redundant. This could be replaced by
```
appendTextToLog(QString("Error parsing file %1").arg(existing_filename))
```
I don't think translating is worth it, since this will only end up in the debug log, which a normal user won't see.

File: mobile-widgets/qmlmanager.cpp
Comment (bstoeger) [reply mobile-widgets/qmlmanager.cpp]: Yes, but the code had the same issue before. Before, it would show more than one message, now it shows the last message. We could accumulate messages in `m_lastError`, but then we could remove the corresponding logic from `report_error()`. That would probably be more robust, since `QString` is thread safe AFAIK. In any case, the fundamental problem of message-mixing remains.

The whole `report_error()`/`get_error_string()` thing is scarily thread unsafe. I think we had a short discussion on the mailing list in which it was concluded with that this should never happen.

I see no problem making it thread safe (without introducing C++isms into the C part), btw. Should I go for it?

**Comment by bstoeger** [comment]
The new version restores the old behavior on mobile: errors are accumulated, which might lead to stale messages being shown.

For the desktop version, I'm not sure if
```
while (!errors.isEmpty())
	getNotificationWidget()->showNotification(errors.takeFirst(), KMessageWidget::Error);
```
will actually show multiple notifications. If not, this will have to be fixed.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [reply mobile-widgets/qmlmanager.cpp]: i think `parse_file()` will report the error that it cannot parse `existing_filename`, then `appendTextToLog()` will report the the error again. i don't think we need `appendTextToLog` here.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [reply mobile-widgets/qmlmanager.cpp]: yep, i know it was the same issue as before. i haven't seen this discussion on the ML, but child threads should simply not call `report_error()`.

if it was already discussed that it should never happen, we can just leave it as is.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: please, use `void qSwap(T & var1, T & var2)`

File: desktop-widgets/mainwindow.cpp
Comment (neolit123) [inline desktop-widgets/mainwindow.cpp]: ~we should probably use `void QString::swap(QString &other)` here.~

not a `QString` it's a `QStringList`. in that case `void qSwap(T & var1, T & var2)` should be used instead of `std::swap()`. we aren't much of a user of the C++ stdlib.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: it would append the same message `parse_file` previous appends via the callback. i think the above call should be removed.

File: desktop-widgets/mainwindow.cpp
Comment (neolit123) [inline desktop-widgets/mainwindow.cpp]: it will just go through the loop and show the last error. this can be seen in `NotificationWidget', so i think there is no point in accumulating them with a mutex. i will make another proposal.

**Comment by neolit123** [comment]
i think we should probably go back the way it was after my first review.
the only fix that was needed was:

> it would append the same message parse_file previous appends via the callback. i think the above call should be removed.

^ it was about `appendTextToLog` being redundant after `parse_file` in `QMLManager::finishSetup()`.

why should be go back?
- it was already discussed on the ML that `report_error` should be used in a single thread e.g. from the UI thread. child threads should handle locks them self if they want to report errors via the same interface.
- our UI does not handle multiple errors notifications - what we can do here is to only show a notification based on the first error, wait for the user to close / discard it and only then show another error, not showing errors which accumulated in the period while the error notification was open. it's not ideal, but the first error is more important.

my comment:
> AFAIK, it's in the same thread, but do we know for sure that m_lastError will be related to loadDivesWithValidCredentials() at all times here and not being an error for something else?

was more of a discussion opener and i didn't want it to spark sudden changes without a discussion.
sorry about that.

**Review by neolit123 — CHANGES_REQUESTED** [review]
please, go back to the way it was after the first review, while applying the mentioned fix in `QMLManager::finishSetup()`. i do not know the state of your local branch but `git reflog` might help here. let me know if you need help with that.

**Comment by bstoeger** [comment]
> it was already discussed on the ML that report_error should be used in a single thread e.g. from the UI thread. 

Note that `report_error()` *is* called from non-UI threads, notably the BT-download thread. That's the reason for the strange code in `MainWindow`. This could perhaps be done more properly with queued signals. So what is the long-term goal here?

File: desktop-widgets/mainwindow.cpp
Comment (neolit123) [inline desktop-widgets/mainwindow.cpp]: we should probably remove this check.

**Comment by neolit123** [comment]
> Note that report_error() is called from non-UI threads, notably the BT-download thread. That's the reason for the strange code in MainWindow. 

i wonder why it isn't a problem then; maybe nothing else is reporting errors at that time?
i would say that the download thread should return a error status rather than calling `report_error()`. that's the fix i would do.

> So what is the long-term goal here?

that's probably a question for @dirkhh.

@bstoeger in terms of the latest implementation, i've posted a couple of comments. my main issue is that our message notifications don't support showing multiple messages. if you keep your `consumeErrors()` idea, i would concat the errors into a single string and show that.

let's see what others have to say about your commits.

**Comment by bstoeger** [comment]
> i wonder why it isn't a problem then; maybe nothing else is reporting errors at that time?

It's hard to imagine how the download thread *and* the UI thread would produce an error at exactly the same time. But if it can be called from different threads, shouldn't it be thread-safe out of principle? I don't know...

**Comment by neolit123** [comment]
> It's hard to imagine how the download thread and the UI thread would produce an error at exactly the same time. But if it can be called from different threads, shouldn't it be thread-safe out of principle? I don't know...

yep, everything can be called from another thread unless there is a restriction for that. my comment from before:
> i would say that the download thread should return a error status rather than calling report_error(). that's the fix i would do.

instead of making `report_error` thread safe, i think we should return states (e.g. integers, statuses) from child threads and then `report_error` again in the main thread. these are my 2c on the subject.

**Comment by dirkhh** [comment]
The idea of report_error() was to make sure we don't have to pass error strings through the deep calling trees. I see no reason why it can't be thread safe - the reason I never bothered to do this was that as indicated above I couldn't envision a scenario where there would be a race.
What I would like to see, if we want to tackle this, is
a) make the function thread safe
b) ensure that when we display the errors, we display all of them
c) make sure that the changes are as minimal as possible

**Comment by bstoeger** [comment]
Ok - next attempt. This uses a `Qt::QueuedConnection` for thread synchronization in the desktop version. The code now looks much cleaner to me.

The mobile version could do something similar, but I'd prefer not to do it, because I can't really test it.

**Review by neolit123 — APPROVED** [review]
lgtm, thanks!

**Comment by bstoeger** [comment]
For consistency, I also made the QMLManager work with a queued connection instead of mutexes. It feels weird to connect an object to itself to transport a string across thread boundaries. Someone with Qt knowledge might want to take a look - is there a more idiomatic/sane way?

PS: mobile only compile tested

**Comment by neolit123** [comment]
@bstoeger you can try:
`QMetaObject::invokeMethod(theMetaObj, "theMethodName", Qt::QueuedConnection, args...);`

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: can be replaced by:
```
QMetaObject::invokeMethod(QMLManager::instance(),
	"showError", Qt::QueuedConnection, Q_ARG(QString, get_error_string()));
```

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: can be removed. no need to connect to self if the `QMetaObject::invokeMethod()` peace is used.

File: mobile-widgets/qmlmanager.h
Comment (neolit123) [inline mobile-widgets/qmlmanager.h]: can be removed if `invokeMethod()` is used.

File: mobile-widgets/qmlmanager.h
Comment (neolit123) [inline mobile-widgets/qmlmanager.h]: should be declared `Q_INVOKABLE void registerError...` if `invokeMethod()` is used.

File: mobile-widgets/qmlmanager.cpp
Comment (neolit123) [inline mobile-widgets/qmlmanager.cpp]: can be replaced by:
```
QMetaObject::invokeMethod(QMLManager::instance(),
	"registerError", Qt::QueuedConnection, Q_ARG(QString, error));
```

Comment (bstoeger) [reply mobile-widgets/qmlmanager.cpp]: Not sure if I like this. :( The whole only-runtime checked / duck typed thing makes me uneasy. Mistype `registerError` or rename the function and you will not get a compile-time error, but subtle behavioral changes. There seems to be a version of `invokeMethod` taking pointer-to-member-function, but only in Qt 5.10+.

So we have the choice between `connect`, which is basically COME FROM and `invokeMethod` which is weakly-typed Python-esque scripting.

Both sound equally unsatisfying to me. :( Any preferences?

Comment (neolit123) [reply mobile-widgets/qmlmanager.cpp]: i'm currently contributing to QtConnectivity where we had the exact same "connect-to-self" situation and one of the maintainers proposed and accepted the `invokeMethod` solution. if a method is renamed `invokeMethod` reports and error and/or crashes (last time i've checked).

**Comment by bstoeger** [comment]
Well, I don't like it, but I also can't change it, so here you go. I changed `Qt::QueuedConnection` to `Qt::AutoConnection`, which only does the queuing thing if called from a different thread.

**Review by neolit123 — APPROVED** [review]
lgtm, thanks for the latest changes.

**Comment by bstoeger** [comment]
If you can test the mobile version, please do. I don't trust it to do the right thing.

**Comment by janmulder** [comment]
Tested on Android device. Did a full new install of the app, synced to one of my test accounts (with lots of data, as it is a copy of my normal logbook), did download from my OSTC3 *bringing in 8 new dives). All seems to be working as before. Inspected the log output of this session, and that looks sane. Not sure it is identical, ie. I did not compare log output from a identical session before and after this PR.

So, lgtm.

**Comment by bstoeger** [comment]
Could you trigger one of the error messages, such as
```
if (save_dives(existing_filename))
```
?
Thanks.

**Comment by janmulder** [comment]
Ok, I created a hacked version to trigger an error (in save_git.c): 

```
if (1 || create_new_commit(repo, remote, branch, &id, create_empty))
		return report_error("creating commit failed");
```
When editing a dive on (android) device, I get a nice toast message `"creating commit failed"` and this message is correctly shown in the log file. Not sure this is what you like to hear, as I do not fully understand the question (ie the reason behind it).

And why a hacked version .... well I do not see any other sane way to generate an error in this execution path.

**Comment by bstoeger** [comment]
> When editing a dive on (android) device, I get a nice toast message "creating commit failed"

Thank you, this is exactly what I hoped to hear. The reason I ask is that the message is transported to the frontend via a `invokeMethod` call. But the way I understand it, the function signature (name + parameters) is only verified at runtime when hitting an error. With `connect` you would at least get an error message on initialization of the `QMLManager` class.

**Pull request merged**
